File: app/src/test/java/me/shreyasr/chatse/network/ClientTest.java
package me.shreyasr.chatse.network;

import android.content.Context;
import android.content.SharedPreferences;

import com.squareup.okhttp.OkHttpClient;
import com.squareup.okhttp.Request;
import com.squareup.okhttp.mockwebserver.MockResponse;
import com.squareup.okhttp.mockwebserver.MockWebServer;

import junit.framework.TestCase;

import java.net.CookieStore;

import me.shreyasr.chatse.TestUtils;
import me.shreyasr.chatse.network.cookie.PersistentCookieStore;

import static org.mockito.Mockito.anyInt;
import static org.mockito.Mockito.anyString;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.when;

public class ClientTest extends TestCase {

    private PersistentCookieStore getCookieStore() {
        SharedPreferences mockSharedPrefs = TestUtils.getMockPrefs();
        Context mockContext = mock(Context.class);

        when(mockContext.getSharedPreferences(anyString(), anyInt())).thenReturn(mockSharedPrefs);

        return new PersistentCookieStore(mockContext);
    }

    public void testClientCustomStore() throws Exception {
        CookieStore cookieStore = getCookieStore();
        OkHttpClient httpClient = new OkHttpClient();

        Client client = new Client(httpClient, cookieStore);
        client.putCookie("http://stackexchange.com", "acct", "content");

        assertEquals("content", client.getCookie("http://stackexchange.com", "acct"));
    }

    public void testClientGetFromWeb() throws Exception {
        MockWebServer server = new MockWebServer();
        server.enqueue(new MockResponse().addHeader("Set-Cookie", "acct=content"));
        server.start();

        CookieStore cookieStore = getCookieStore();
        OkHttpClient httpClient = new OkHttpClient();

        Request request = new Request.Builder()
                .url(server.getUrl("/"))
                .build();

        Client client = new Client(httpClient, cookieStore);
        client.newCall(request).execute();

        assertEquals("content", client.getCookie(server.getUrl("/").toURI(), "acct"));
    }
}

File: app/src/test/java/com/tristanwiley/chatse/network/ClientTest.kt
package com.tristanwiley.chatse.network

import android.content.Context
import com.nhaarman.mockitokotlin2.any
import com.nhaarman.mockitokotlin2.doReturn
import com.nhaarman.mockitokotlin2.mock
import com.nhaarman.mockitokotlin2.whenever
import com.squareup.okhttp.OkHttpClient
import com.squareup.okhttp.Request
import com.squareup.okhttp.mockwebserver.MockResponse
import com.squareup.okhttp.mockwebserver.MockWebServer
import com.tristanwiley.chatse.getMockPrefs
import com.tristanwiley.chatse.network.cookie.PersistentCookieStore
import org.junit.Test
import kotlin.test.assertEquals

class ClientTest {

    @Test
    fun testClientCustomStore() {
        val cookieStore = givenCookieStore()
        val httpClient = OkHttpClient()
        val client = Client(httpClient, cookieStore)

        client.putCookie("http://stackexchange.com", "acct", "content")

        val expected = "content"
        val actual = client.getCookie("http://stackexchange.com", "acct")
        assertEquals(expected, actual)
    }

    @Test
    fun testClientGetFromWeb() {
        val server = MockWebServer()
        server.enqueue(MockResponse().addHeader("Set-Cookie", "acct=content"))
        server.start()

        val cookieStore = givenCookieStore()
        val httpClient = OkHttpClient()
        val client = Client(httpClient, cookieStore)

        val request = Request.Builder().apply {
            url(server.getUrl("/"))
        }.build()

        client.newCall(request).execute()

        val expected = "content"
        val actual = client.getCookie(server.getUrl("/").toURI(), "acct")
        assertEquals(expected, actual)

    }

    private fun givenCookieStore(): PersistentCookieStore {
        val sharedPefs = getMockPrefs()
        val context = mock<Context>()

        whenever(context.getSharedPreferences(any(), any())).doReturn(sharedPefs)

        return PersistentCookieStore(context)
    }

}
